### PR TITLE
[Document Provider] Edit files and save changes

### PR DESCRIPTION
**Pull request by abelgardep**
Implements #2498 

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/Android/Release_2.11.0/2527-document_provider_edit_file.md

Bugs & improvements:

- [x] (1) Behaviour with no connection on non-downloaded files https://github.com/owncloud/android/pull/2527#issuecomment-489058679 [WONT FIX]

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 1]
```diff
@@ -87,7 +92,40 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
+        val accessMode: Int = ParcelFileDescriptor.parseMode(mode)
+
+        val isWrite: Boolean = mode.contains("w")
+        return if (isWrite) {
+            val handler = Handler(context?.mainLooper)
+            // Attach a close listener if the document is opened in write mode.
+            try {
+                ParcelFileDescriptor.open(File(file.storagePath), accessMode, handler) {
+                    // Update the file with the cloud server. The client is done writing.
+                    Log_OC.d(TAG, "A file with id $documentId has been closed! Time to update the server.")
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: I would say "time to synchronize it with server"

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 2]
```diff
@@ -87,7 +92,40 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
+                            }
+                        }
+                    }.apply { start() }
+                }
+            } catch (e: IOException) {
+                throw FileNotFoundException(
+                    "Failed to open document with id $documentId and mode $mode"
+                )
+            }
+        } else {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: You could avoid this `else` condition by doing something like 

```
if (!isWrite) return ParcelFileDescriptor.open(File(file.storagePath), accessMode)

// Attach a close listener if the document is opened in write mode.
try {
    val handler = Handler(context?.mainLooper)
    
    return ParcelFileDescriptor.open(File(file.storagePath), accessMode, handler) {
        // Update the file with the cloud server. The client is done writing.
        Log_OC.d(TAG, "A file with id $documentId has been closed! Time to update the server.")
        Thread {
            SynchronizeFileOperation(
                file,
                null,
                currentStorageManager?.account,
                false,
                context,
                false
            ).apply {
                val result = execute(currentStorageManager, context)
                if (result.code == RemoteOperationResult.ResultCode.SYNC_CONFLICT) {
                    NotificationUtils.notifyConflict(file, currentStorageManager?.account, context)
                }
            }
        }.apply { start() }
    }
} catch (e: IOException) {
    throw FileNotFoundException(
        "Failed to open document with id $documentId and mode $mode"
    )
}
```

**Comment by davigonz** [comment]
@abelgardep just a couple of changes...

**Comment by abelgardep** [comment]
@davigonz Changes applied

**Comment by davigonz** [comment]
Code approved, ready to test @jesmrec

**Comment by jesmrec** [comment]
Starting QA...

**Comment by jesmrec** [comment]
### (1)

Check these steps:

1. Remove connection from device
2. From downloads/files app, try to open a non-downloaded file

Current behaviour: file is tried to be opened forever
Expected: i reproduced the same steps over `master` branch. There, when the file is retrieved for the 3rd party app, an error message is displayed

Notification is OK, showing the connection problem

With `master`: 

May-03-2019 12-46-39

With current branch:

May-03-2019 12-56-53

Nexus 6P v7

**Comment by jesmrec** [comment]
(1) is finally reproducible in master, so it will be moved to a separate issue. 

apart of that, everything OK

Approved

**Pull request merged**